bn-apps/memberships-management/membership-service/src/main/kotlin/com/r3/businessnetworks/membership/flows/bno/service/DatabaseService.kt

```diff
@@ -1,100 +1,106 @@
+ */
+@CordaService
+class DatabaseService(val serviceHub: ServiceHub) : SingletonSerializeAsToken() {
+    fun getMembership(member: Party, bno: Party): StateAndRef<MembershipState<Any>>? {
+        val criteria = QueryCriteria.VaultQueryCriteria(Vault.StateStatus.UNCONSUMED)
+                .and(memberCriteria(member))
+                .and(bnoCriteria(bno))
+        val states = serviceHub.vaultService.queryBy<MembershipState<Any>>(criteria).states
+        return if (states.isEmpty()) null else (states.sortedBy { it.state.data.modified }.last())
+    }
+
+    /**
+     * This method is used to search for memberships on a particular network since the BNO can be in control on more then one network and the member can exsit on multiple networks
+     */
+    fun getMembershipOnNetwork(member: Party, bno: Party, networkID: String?): StateAndRef<MembershipState<Any>>? {
+        val criteria = QueryCriteria.VaultQueryCriteria(Vault.StateStatus.UNCONSUMED)
+                .and(memberCriteria(member))
+                .and(networkIDCriteria(networkID))
+                .and(bnoCriteria(bno))
+        val states = serviceHub.vaultService.queryBy<MembershipState<Any>>(criteria).states
+        return if (states.isEmpty()) null else (states.sortedBy { it.state.data.modified }.last())
+    }
+
+    fun getAllMemberships(bno: Party): List<StateAndRef<MembershipState<Any>>> {
+        val criteria = QueryCriteria.VaultQueryCriteria(Vault.StateStatus.UNCONSUMED)
+                .and(bnoCriteria(bno))
+        return serviceHub.vaultService.queryBy<MembershipState<Any>>(criteria).states
+    }
+
+    fun getActiveMemberships(bno: Party) = getAllMemberships(bno).filter { it.state.data.isActive() }
+
+    /**
+     * This method exists to prevent the same member from being able to request a membership multiple times, while their first membership transaction is being processed.
+     * Pending membership request is created when a membership request arrives and is deleted when the membership transaction is finalised.
+     * Any attempt to create a duplicate pending membership request would violate the DB constraint.
+     */
+    fun createPendingMembershipRequest(party: Party) {
+        serviceHub.withEntityManager {
+            persist(PersistentPendingMembershipRequest.from(party))
+            flush()
+        }
+    }
+
+    fun deletePendingMembershipRequest(party: Party) {
+        serviceHub.withEntityManager {
+            val hqlQuery = """
+                delete from PersistentPendingMembershipRequest
+                where pendingMember = :pendingMember
+            """
+
+            createQuery(hqlQuery)
+                    .setParameter("pendingMember", party.name.toString())
+                    .executeUpdate()
+        }
+    }
+
+
+    private fun networkIDCriteria(NetworkID: String?) = QueryCriteria.VaultCustomQueryCriteria(builder { MembershipStateSchemaV1.PersistentMembershipState::networkID.equal(NetworkID) })

```

Comment:
Do we know if this query is safe with `NetworkID=null`?

Reply:
NetworkID isn't safe with null value when being used for database query the test fails, currently working on a way to allow the Database to be queried with null value